### PR TITLE
Misc improvements

### PR DESCRIPTION
**Pull request by elicn**
A bunch of non-related commits with small improvements.

Changes highlights:
- Modify some tiny aspects of memory API
- Heavily annotate and comment core_hooks

Hooks are usually the first functionality new users try out, and also struggle to understand.
Hopefully the documentation and annotations there would help them find their way there.

### CODE REVIEW

File: qiling/core_hooks.py [hunk 1]
```diff
@@ -18,6 +19,73 @@
 from .const import QL_HOOK_BLOCK
 from .exception import QlErrorCoreHook
 
+if TYPE_CHECKING:
```
Comment (wtdcode) [inline qiling/core_hooks.py:22]: try `from __future__ import annotations`, which is more elegant.

Comment (elicn) [reply qiling/core_hooks.py:22]: importing `annotations` helps with avoiding string annotations for circular dependencies (i.e. no need to to do `ql: 'Qiling'` anymore, just `ql: Qiling`), but it still requires the `TYPE_CHECKING` mechanism.

Is this what you meant..?

File: qiling/os/memory.py [hunk 2]
```diff
@@ -166,21 +166,21 @@ def __perms_mapping(ps: int) -> str:
...
-                container = None
+                container = ''
 
             return (lbound, ubound, perms_str, label, container)
 
         return tuple(__process(*entry) for entry in self.map_info)
 
-    def show_mapinfo(self):
-        """Emit memory map info in a nicely formatted table.
+    def get_formatted_mapinfo(self) -> Sequence[str]:
```
[the remark this answers is not in the document]
Comment (elicn) [reply qiling/os/memory.py:182]: The method behvaior changed from (emitting) the information using `ql.log.info` to returning a list of strings ready to print. That gives the user the flexibility to choose how to print it, including using another logging level. There were two main reasons for that:
1. It was used mainly on crashes, and it looked wierd that the entire crash info was emitted with `ql.log.error` while this one was printed with another logging level
2. Embedding side-effects in functions is usually not a good idea, and printing is a good example for such side-effect. I rather let the user do the printing

So to your question, this method no longer "shows", it just "formats" the info and returns it to the user to do the printing.

[the remark this answers is not in the document]
Comment (kabeor) [reply qiling/os/memory.py:182]: Well, I mean we can stay `show_mapinfo()` in utils.py or somewhere, so users needn't to code a `for loop` to print by themselves. Their exist scripts should be still compatible.

[the remark this answers is not in the document]
Comment (elicn) [reply qiling/os/memory.py:182]: I don't know how much it was used outside of Qiling crash, but I guess we could add that somewhere..

**Review by kabeor — APPROVED** [review]

**Review by cla7aye15I4nd — APPROVED** [review]

File: qiling/core_hooks.py (hunk 1 above)
Comment (wtdcode) [reply qiling/core_hooks.py:22]: > importing `annotations` helps with avoiding string annotations for circular dependencies (i.e. no need to to do `ql: 'Qiling'` anymore, just `ql: Qiling`), but it still requires the `TYPE_CHECKING` mechanism.
> 
> Is this what you meant..?

Sorry I missed the notification. Yes you are right, for circular dep, `TYPE_CHECKING` is the only way.

**Review by wtdcode — APPROVED** [review]

**Pull request merged**